### DISCUSSION TITLE
dwd-mosmix, --belchertown, AQI Daten

### DISCUSSION THREAD

**Discussion by hoetzgit** [roe-dl/weewx-DWD#10] [Ideas]
Hallo, wir haben uns ja kürzlich bereits über das Thema AQI und das Fehlen in der `forecast.json` ausgetauscht.

Für mich habe ich das jetzt temporär so gelöst:
- vor `dwd-mosmix` per cron einen Airquality API Aufruf bei AerisWeather. Der output wird als JSON-Datei in das per `path` konfigurierte Verzeichnis gespeichert.
- `dwd-mosmix` schaut, ob eine AQI JSON Datei vorhanden ist (analog Warnungen) , wenn ja, wird der Inhalt in den aqi Teil der `forecast.json` eingebunden, sonst wie gehabt leer.

Wäre es sinnvoll dieses Prinzip (o.ä.) in `dwd-mosmix` für `--belchertown` zu übernehmen? Bspw. mit zus. optionalen Parameter `aqi_forecast_file` in der `[[Belchertown]]` Konfiguration?

Mir ist allerdings auch klar, dass dies dann nichts mit den Daten des DWD zu tun hat.

**Comment by roe-dl** [comment]
Es ist eine Idee. Ich sehe es mir mal an.

**Comment by roe-dl** [comment]
Deutsche Meßdaten bekommt man beim Umweltbundesamt. Die URL ist:
```
https://www.umweltbundesamt.de/api/air_data/v2/airquality/csv?date_from=2022-02-21&time_from=1&date_to=2022-03-01&time_to=1&station=1589&lang=de01
```
Datum und Stationsnummer muß man entsprechend anpassen. Das Ergebnis sind Meßwerte für Stickoxid, Feinstaub und Ozon sowie eine qualitative Bewertung in fünf Stufen in englischer Sprache. Die Stationen befinden sich überwiegend in größeren Städten. In ländlicher Gegend gibt es so gut wie gar keine.

Link: [Umweltbundesamt](https://www.umweltbundesamt.de/daten/luft/luftdaten/luftqualitaet/eJzrWJSSuMrIwMhI18BI18hiUUnmIkPzRXmpiwyNFxWXLDE0tbBcnOJWhFBiuDglJB9ZR24Vx6Lc5KbFOYklpx08C_J4lTxzF-fkpZ920Ko3mvry4TIAwbMkiw==)

**Comment by roe-dl** [comment]
@hoetzgit Ich habe den Fork wahrgenommen. Soll es da einen Pull-Request zu diesem Punkt geben?

**Reply by hoetzgit** [reply]
Nur wenn es sinnvoll erscheint. Hab's jetzt für mich mal so gemacht um auch die Datei `forecast.json` wieder komplett zu haben und damit alle Parameter der Belchertown Skin nutzen zu können. Mit den AQ Daten haben wir ja schon gesprochen, dass das so eine Sache ist.
Passt das überhaupt so? Habe von Python keine Ahnung.

**Reply by roe-dl** [reply]
Für "von Python keine Ahnung" sehen die Änderungen aber ziemlich professionell aus. Ich hatte auch noch keine Zeile Python geschrieben, bevor ich auf WeeWX gekommen bin. Learning by Doing.

**Reply by hoetzgit** [reply]
Hab auch von Github keine Ahnung, also nicht wundern ;-)

**Reply by roe-dl** [reply]
Ich habe schmerzhaft gelernt, daß es wichtig ist, für jeden Pull-Request vorher bei sich einen eigenen Tag anzulegen, sonst reagiert github etwas komisch.

**Reply by hoetzgit** [reply]
Ich habe die Anpassungen im Branch `PWS-Weiherhammer` den AQ Teil in `dwd-mosmix` für die AerisWeather Daten abgeschlossen. Zusätzlich habe ich den Output der compass Funktion in die forecast.json hinzugefügt und nutze das in der Vorhersage in der Belchertown Skin. Da es zusätzliche Daten sind, denke ich, dass das unkritisch ist. Vielleicht können Sie (Du?) ja mal bei Gelegenheit drüber schauen, ob die Anpassungen in `dwd-mosmix` so passen.

**Reply by c0by85** [reply]
Kurze Frage nebenbei, würde das den Fehler beheben / erklären?

```
message repeated 15 times: [ ERROR user.belchertown: Error getting AQI from Aeris weather. The error was: string indices must be integers]
ERROR user.belchertown: Error getting AQI from Aeris weather. The error was: string indices must be integers
```

**Reply by hoetzgit** [reply]
Ich glaube fast, dass war der Grund warum ich die Anpassungen gemacht habe und den Aeris AQ Teil ergänzen wollte. Hatte auch den Parameter `aqi_enabled`/`aqi_location_enabled` in der Belchertown Skin gesetzt und dann kam ursprünglich dieser Fehler.

**Reply by c0by85** [reply]
Hatte eigentlich geguckt und zu AQI nichts weitergefunden, dachte es wäre aus gewesen. Aber mit der Änderung aus `PWS-Weiherhammer`, gibt es den Fehler nicht mehr?! Mal probieren, oder nach der anderen Option suchen.

Genau, beides ist aus und trotzdem der Fehler.

```
aqi_enabled = 0
aqi_location_enabled = 0
```

**Reply by hoetzgit** [reply]
Auf meinem WeeWX Backup habe ich den Fehler aktuell auch noch, nach den Anpassungen auf meinem "Produktivsystem" nicht mehr. Bei mir sind beide Parameter aktiv und der AQI inklusive Quelle wird auf der Hauptseite angezeigt. Der `[aqi]` Teil der `forecast.json` ist nun identisch zu den bisherigen  API Aufrufen aus der Belchertown Skin (falls ich nichts übersehen habe).

**Reply by hoetzgit** [reply]
Auf die Schnelle mal geschaut, evtl führt die Abfrage

```
elif (
                    data["aqi"][0]["error"]["code"] == "warn_no_data"
                ):

```
in der `belchertown.py` zu diesem Fehler. Ein Element `["code"]` ist nicht vorhanden und führt zu einer Exception. Dann müsste dies nur in `dwd-mosmix` ergänzt werden.

Beispiel eines gültigen `'aqi'` JSON outputs im Fehlerfall:

```
{
    "success": false,
    "error": {
        "code": "invalid_location",
        "description": "The requested location was not found."
    },
    "response": []
}

```
Du könntest folgende Anpassung in `dwd-mosmix` Zeile `1352` vornehmen und testen:

`'error':'not available',`
zu
`'error':{'code':'not available','description':'not available'},`

**Reply by roe-dl** [reply]
> Ich habe die Anpassungen im Branch `PWS-Weiherhammer` den AQ Teil in `dwd-mosmix` für die AerisWeather Daten abgeschlossen. Zusätzlich habe ich den Output der compass Funktion in die forecast.json hinzugefügt und nutze das in der Vorhersage in der Belchertown Skin. Da es zusätzliche Daten sind, denke ich, dass das unkritisch ist. Vielleicht können Sie (Du?) ja mal bei Gelegenheit drüber schauen, ob die Anpassungen in `dwd-mosmix` so passen.

 * 'Du' ist OK
 * Ich habe es mir den Punkt mit der compass()-Funktion angesehen. Das paßt so, denke ich. 
 * Der originale Schlüssel für die Windrichtung als Himmelsrichtung (dort ohne Gradzahl) bei Aeris ist schlicht `windDir`.

**Reply by hoetzgit** [reply]
Vielen Dank, dann werde ich morgen mal mergen üben 😏

**Comment by hoetzgit** [comment]
Vielen Dank für die AerisWeather API Anpassung.

Nur ein Gedanke: Man sollte noch das AerisWeather API Limit (beim PWS Contributors Package) im Auge behalten, dazu vielleicht erstmal ein Hinweis in der readme? Auf längere Sicht evtl. eine zeitliche Prüfung - wie in der Skin - und dann unter Verwendung der "Alt"-Daten. Obwohl - 5000 Aufrufe muss man auch erstmal schaffen ;-)

**Reply by roe-dl** [reply]
Ich würde erst einmal dahin orientieren, die Daten direkt beim Umweltbundesamt herunterzuladen. Von dort kann sie Aeris auch nur haben, und dort haben wir kein solches Limit. Ein Hinweis in der README-Datei schadet natürlich trotzdem nicht.

**Reply by hoetzgit** [reply]
Habe mal geschaut, Aeris gibt bei mir [OpenAQ](https://openaq.org/#/) als Quelle an. Da meine nächste Station nur Feinstaub, Stickstoffdioxid und Ozon misst (Beim UBA sind auch nur diese gelistet), ist für mich immer unklar woher der Rest kommt. Aber mein SDS011 Sensor ist da, da messe ich dann den Feinstaub selbst und hab wieder was zum basteln.

**Reply by hoetzgit** [reply]
Auch eine mögliche Quelle: https://public.opendatasoft.com/explore/dataset/api-luftdateninfo/information/
